// http-aws-es/legacy.js
'use strict';

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _awsSdk = require('aws-sdk');

var _awsSdk2 = _interopRequireDefault(_awsSdk);

var _http = require('elasticsearch/src/lib/connectors/http');

var _http2 = _interopRequireDefault(_http);

var _utils = require('elasticsearch/src/lib/utils');

var _utils2 = _interopRequireDefault(_utils);

var _zlib = require('zlib');

var _zlib2 = _interopRequireDefault(_zlib);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * A connection handler for Amazon ES.
 *
 * Uses the aws-sdk to make signed requests to an Amazon ES endpoint.
 *
 * @param client {Client} - The Client that this class belongs to
 * @param config {Object} - Configuration options
 * @param [config.protocol=http:] {String} - The HTTP protocol that this connection will use, can be set to https:
 * @class HttpConnector
 */

var HttpAmazonESConnector = function (_HttpConnector) {
  (0, _inherits3.default)(HttpAmazonESConnector, _HttpConnector);

  function HttpAmazonESConnector(host, config) {
    (0, _classCallCheck3.default)(this, HttpAmazonESConnector);

    var _this = (0, _possibleConstructorReturn3.default)(this, (HttpAmazonESConnector.__proto__ || (0, _getPrototypeOf2.default)(HttpAmazonESConnector)).call(this, host, config));

    var protocol = host.protocol,
        port = host.port;

    var endpoint = new _awsSdk2.default.Endpoint(host.host);

    // #10
    if (protocol) endpoint.protocol = protocol.replace(/:?$/, ":");
    if (port) endpoint.port = port;

    _this.AWS = _awsSdk2.default;
    _this.endpoint = endpoint;
    return _this;
  }

  (0, _createClass3.default)(HttpAmazonESConnector, [{
    key: 'request',
    value: function () {
      var _ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee(params, cb) {
        var incoming, timeoutId, request, req, status, headers, log, response, AWS, reqParams, cleanUp, p, CREDS, signer, send;
        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                incoming = void 0;
                timeoutId = void 0;
                request = void 0;
                req = void 0;
                status = 0;
                headers = {};
                log = this.log;
                response = void 0;
                AWS = this.AWS;
                reqParams = this.makeReqParams(params);
                // general clean-up procedure to run after the request
                // completes, has an error, or is aborted.

                cleanUp = _utils2.default.bind(function (err) {
                  clearTimeout(timeoutId);

                  req && req.removeAllListeners();
                  incoming && incoming.removeAllListeners();

                  if (err instanceof Error === false) {
                    err = void 0;
                  }

                  log.trace(params.method, reqParams, params.body, response, status);
                  if (err) {
                    cb(err);
                  } else {
                    cb(err, response, status, headers);
                  }
                }, this);


                request = new AWS.HttpRequest(this.endpoint);

                // copy across params
                for (p in reqParams) {
                  request[p] = reqParams[p];
                }
                request.region = AWS.config.region;
                if (params.body) request.body = params.body;
                if (!request.headers) request.headers = {};
                request.headers['presigned-expires'] = false;
                request.headers['Host'] = this.endpoint.host;

                // load creds
                // #1, #3, #12, #15, #16, #21
                _context.next = 20;
                return this.getAWSCredentials();

              case 20:
                CREDS = _context.sent;


                // Sign the request (Sigv4)
                signer = new AWS.Signers.V4(request, 'es');

                signer.addAuthorization(CREDS, new Date());

                send = new AWS.NodeHttpClient();

                req = send.handleRequest(request, null, function (_incoming) {
                  incoming = _incoming;
                  status = incoming.statusCode;
                  headers = incoming.headers;
                  response = '';

                  var encoding = (headers['content-encoding'] || '').toLowerCase();
                  if (encoding === 'gzip' || encoding === 'deflate') {
                    incoming = incoming.pipe(_zlib2.default.createUnzip());
                  }

                  incoming.setEncoding('utf8');
                  incoming.on('data', function (d) {
                    response += d;
                  });

                  incoming.on('error', cleanUp);
                  incoming.on('end', cleanUp);
                }, cleanUp);

                req.on('error', cleanUp);

                req.setNoDelay(true);
                req.setSocketKeepAlive(true);

                return _context.abrupt('return', function () {
                  req.abort();
                });

              case 29:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function request(_x, _x2) {
        return _ref.apply(this, arguments);
      }

      return request;
    }()
  }, {
    key: 'getAWSCredentials',
    value: function getAWSCredentials() {
      var AWS = this.AWS;


      return new _promise2.default(function (resolve, reject) {
        AWS.config.getCredentials(function (err, creds) {
          if (err) return reject(err);
          return resolve(creds);
        });
      });
    }
  }]);
  return HttpAmazonESConnector;
}(_http2.default);

module.exports = HttpAmazonESConnector;
